src/content/developers/docs/consensus-mechanisms/pow/mining-algorithms/dagger-hashamoto/index.md

```diff
@@ -0,0 +1,327 @@
+
+Essentially, it starts off a graph as a single node, `sha3(seed)`, and from there starts sequentially adding on other nodes based on random previous nodes. When a new node is created, a modular power of the seed is computed to randomly select some indices less than `i` (using `x % i` above), and the values of the nodes at those indices are used in a calculation to generate a new a value for `x`, which is then fed into a small proof of work function (based on XOR) to ultimately generate the value of the graph at index `i`. The rationale behind this particular design is to force sequential access of the DAG; the next value of the DAG that will be accessed cannot be determined until the current value is known. Finally, modular exponentiation is used to further hash the result.
+
+This algorithm relies on several results from number theory. See the appendix below for a discussion.
+
+## Light Client Evaluation
+
+The intent of the above graph construction is to allow each individual node in the graph can be reconstructed by computing a subtree of only a small number of nodes, and requiring only a small amount of auxiliary memory. Note that with k=1, the subtree is only a chain of values going up to the first element in the DAG.
+
+The light client computing function for the DAG works as follows:
+
+```python
+def quick_calc(params, seed, p):
+    w, P = params["w"], params["P"]
+    cache = {}
+
+    def quick_calc_cached(p):
+        if p in cache:
+            pass
+        elif p == 0:
+            cache[p] = pow(sha3(seed), w, P)
+        else:
+            x = pow(sha3(seed), (p + 1) * w, P)
+            for _ in range(params["k"]):
+                x ^= quick_calc_cached(x % p)
+            cache[p] = pow(x, w, P)
+        return cache[p]
+
+    return quick_calc_cached(p)
+```
+
+Essentially, it is simply a rewrite of the above algorithm that removes the loop of computing the values for the entire DAG and replaces the earlier node lookup with a recursive call or a cache lookup. Note that for `k=1` the cache is unnecessary, although a further optimization actually precomputes the first few thousand values of the DAG and keeps that as a static cache for computations; see the appendix for a code implementation of this.
+
+## Double Buffer of DAGs
+
+In a full client, a [_double buffer_](https://en.wikipedia.org/wiki/Multiple_buffering) of 2 DAGs produced by the above formula is used. The idea is that DAGs are produced every `epochtime` number of blocks according to the params above. The client does not use the latest DAG produced, but the previous one. The benefit of this is that it allows the DAGs to be replaced over time without needing to incorporate a step where miners must suddenly recompute all of the data. Otherwise, there is the potential for an abrupt temporary slowdown in chain processing at regular intervals and dramatically increasing centralization and thus 51% attack risks within those few minutes before all data is recomputed.
```

Comment:
```suggestion
In a full client, a [_double buffer_](https://wikipedia.org/wiki/Multiple_buffering) of 2 DAGs produced by the above formula is used. The idea is that DAGs are produced every `epochtime` number of blocks according to the params above. Instead of the client using the latest DAG produced, it uses the one previous. The benefit of this is that it allows the DAGs to be replaced over time without needing to incorporate a step where miners must suddenly recompute all of the data. Otherwise, there is the potential for an abrupt temporary slowdown in chain processing at regular intervals and dramatically increasing centralization. Thus 51% attack risks within those few minutes before all data are recomputed.
```